server/sonar-server/src/main/java/org/sonar/server/computation/source/DuplicationLineReader.java

```diff
@@ -20,93 +20,121 @@
 
-  public DuplicationLineReader(Iterator<BatchReport.Duplication> duplications) {
-    this.duplications = newArrayList(duplications);
-    // Sort duplication to have deterministic results and avoid false variation that would lead to an unnecessary update of the source files
-    // data
-    Collections.sort(this.duplications, new DuplicationComparator());
-
-    this.duplicationIdsByRange = createDuplicationIdsByRange(this.duplications);
+  public DuplicationLineReader(Set<Duplication> duplications) {
+    this.duplicatedTextBlockIndexByTextBlock = createIndexOfDuplicatedTextBlocks(duplications);
   }
 
   @Override
   public void read(DbFileSources.Line.Builder lineBuilder) {
-    int line = lineBuilder.getLine();
-    List<BatchReport.TextRange> blocks = findDuplicationBlockMatchingLine(line);
-    for (BatchReport.TextRange block : blocks) {
-      lineBuilder.addDuplication(duplicationIdsByRange.get(block));
+    Predicate<Map.Entry<TextBlock, Integer>> containsLine = new TextBlockContainsLine(lineBuilder.getLine());
+    for (Integer textBlockIndex : from(duplicatedTextBlockIndexByTextBlock.entrySet())
+        .filter(containsLine)
+        .transform(MapEntryToBlockId.INSTANCE)
+            // list is sorted to cope with the non-guaranteed order of Map entries which would trigger false detection of changes
+            // in {@link DbFileSources.Line#getDuplicationList()}
+        .toSortedList(Ordering.natural())) {
+      lineBuilder.addDuplication(textBlockIndex);
     }
   }
 
-  private List<BatchReport.TextRange> findDuplicationBlockMatchingLine(int line) {
-    List<BatchReport.TextRange> blocks = newArrayList();
-    for (BatchReport.Duplication duplication : duplications) {
-      if (matchLine(duplication.getOriginPosition(), line)) {
-        blocks.add(duplication.getOriginPosition());
-      }
-      for (BatchReport.Duplicate duplicate : duplication.getDuplicateList()) {
-        if (isDuplicationOnSameFile(duplicate) && matchLine(duplicate.getRange(), line)) {
-          blocks.add(duplicate.getRange());
-        }
-      }
-    }
-    return blocks;
+  private static boolean isLineInBlock(TextBlock range, int line) {
+    return line >= range.getStart() && line <= range.getEnd();
   }
 
-  private static boolean isDuplicationOnSameFile(BatchReport.Duplicate duplicate) {
-    return !duplicate.hasOtherFileRef();
-  }
+  /**
+   *
+   * <p>
+   * This method uses the natural order of TextBlocks to ensure that given the same set of TextBlocks, they get the same
+   * index. It avoids false detections of changes in {@link DbFileSources.Line#getDuplicationList()}.
+   * </p>
+   */
+  private static Map<TextBlock, Integer> createIndexOfDuplicatedTextBlocks(Collection<Duplication> duplications) {
+    List<TextBlock> duplicatedTextBlocks = extractAllDuplicatedTextBlocks(duplications);
+    Collections.sort(duplicatedTextBlocks);
+    return from(duplicatedTextBlocks)
+        .toMap(new Function<TextBlock, Integer>() {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make this a named "static" inner class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2694)